Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -26,51 +26,368 @@
+#include "TextBoundaries.h"
 #include "TextIterator.h"
 
+#include <dbg_editing.h>
+#include <dbg_dom.h>
+
 namespace WebCore {
 namespace FragmentDirectiveRangeFinder {
 
-static SimpleRange collapseIfRootsDiffer(SimpleRange&& range)
+static bool isSearchInvisible(Node& node)
 {
-    return &range.start.container->rootNode() == &range.end.container->rootNode()
-        ? WTFMove(range) : SimpleRange { range.start, range.start };
+    // TODO: The computed value of its display property is none.
+    
+    // TODO: If the node serializes as void.
+    
+    if (is<HTMLIFrameElement>(node) || is<HTMLImageElement>(node) || is<HTMLMeterElement>(node) || is<HTMLObjectElement>(node) || is<HTMLProgressElement>(node) || is<HTMLStyleElement>(node) || is<HTMLScriptElement>(node) || is<HTMLVideoElement>(node) || is<HTMLAudioElement>(node))
+        return true;
+    
+    // TODO: Is a select element whose multiple content attribute is absent.
+    
+    return false;
 }
 
-static std::optional<SimpleRange> rangeOfString(const String& target, Document& document)
+static bool isNonSearchableSubtree(Node& node)
 {
-    if (target.isEmpty())
+    Node* traversingNode = &node;
+    while(traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentInComposedTree();
+    }
+    return false;
+}
+
+static Node& nearestBlockAncestor(Node& node)
+{
+    Node* curNode = &node;
+    while (curNode)
+    {
+        if (!curNode->isTextNode() && curNode->renderer() && curNode->renderer()->style().isDisplayBlockLevel())
+            return *curNode;
+        else
+            curNode = curNode->parentInComposedTree();
+    }
+    return node.document();
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<Node*> textNodeList, bool isEnd)
+{
+    unsigned counted = 0;
+    
+    for (auto* curNode : textNodeList) {
+        unsigned nodeEnd = counted + curNode->length();
+        if (isEnd)
+            nodeEnd++;
+        if (nodeEnd > index)
+            return BoundaryPoint(*curNode, index - counted);
+        counted += curNode->length();
+    }
+    return std::nullopt;
+}
+
+static bool indexIsWordBoundary(const String& string, unsigned index)
+{
+    UBreakIterator* it = wordBreakIterator(string);
+    return ubrk_isBoundary(it, index);
+}
+
+static bool isVisibleTextNode(Node& node)
+{
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange searchRange, Vector<Node*> nodes, bool wordStartBounded, bool wordEndBounded)
+{
+    String searchBuffer;
+    std::optional<BoundaryPoint> start;
+    std::optional<BoundaryPoint> end;
+    
+    if (nodes.size() < 1)
         return std::nullopt;
+    
+    StringBuilder searchBufferBuilder;
+    for (auto* node : nodes) {
+        if (node)
```

Comment:
Why this if check, I don't think you can ever put null in your Vector? Also, in the other place where you iterate over the vector, you don't null check.

As a side note, we should make sure that nothing can destroy the nodes while they in your vector since you're only holding raw pointers. It looks dangerous but hopefully safe here?